In [2]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import numpy as np
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [3]:
# Load the dataset (load remotely or locally)
iris_dataset = pd.read_csv("iris.csv")

# Output the first 15 rows of the data
iris_dataset.head(15)

# Display a summary of the table information (number of datapoints, etc.)
#iris_dataset.info()


,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa
5,5.4,3.9,1.7,0.4,setosa
6,4.6,3.4,1.4,0.3,setosa
7,5.0,3.4,1.5,0.2,setosa
8,4.4,2.9,1.4,0.2,setosa
9,4.9,3.1,1.5,0.1,setosa


## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

Each row of the data represents the characteristics of each flower found in the iris dataset. The features of the dataset are the sepal_length, sepal_width, petal_length, petal_width and the label is the species. Upon looking up the scikit documentation of the dataset, there are three possible classes for each row: setosa, versicolor, and virginica. Each of the labels corresponds to the distinct species of the iris flower (Iris setosa, Iris versicolor and Iris virginica)

# Part 2: Split the dataset into train and test

In [4]:
# Take the dataset and split it into our features (X) and label (y)
X = iris_dataset.iloc[:,:-1] # our feature columns (as a Dataframe)
y = iris_dataset.iloc[:,-1] # our label columns (as a Series)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=41)
X_test

,sepal_length,sepal_width,petal_length,petal_width
119,6.0,2.2,5.0,1.5
128,6.4,2.8,5.6,2.1
135,7.7,3.0,6.1,2.3
91,6.1,3.0,4.6,1.4
112,6.8,3.0,5.5,2.1
71,6.1,2.8,4.0,1.3
123,6.3,2.7,4.9,1.8
85,6.0,3.4,4.5,1.6
147,6.5,3.0,5.2,2.0
143,6.8,3.2,5.9,2.3


# Part 3: Logistic Regression

In [5]:
# i. Use sklearn to train a LogisticRegression model on the training set
clf_logreg = LogisticRegression().fit(X=X_train, y=y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print("Probabilities for a random datapoint: ", clf_logreg.predict_proba(X_test.sample())) # returns [[9.81141289e-01 1.88586775e-02 3.30699410e-08]], for setosa, versicolor, virginica respectively

# iii. Report on the score for Logistic regression model, what does the score measure?
print("Score:", clf_logreg.score(X_test, y_test))
print("This score measures the mean accuracy on the given train data and labels. A number closer to 1 is more accurate.")

# iv. Extract the coefficents and intercepts for the boundary line(s)
print("Boundary line coefficients are", clf_logreg.coef_)
print("Boundary line intercepts are", clf_logreg.intercept_)


Probabilities for a random datapoint:  [[7.13117249e-07 1.75528582e-02 9.82446429e-01]]
Score: 0.9333333333333333
This score measures the mean accuracy on the given train data and labels. A number closer to 1 is more accurate.
Boundary line coefficients are [[-0.43525979  0.87576952 -2.4200923  -1.04855271]
 [ 0.56849391 -0.4854157  -0.19972697 -0.8797814 ]
 [-0.13323412 -0.39035382  2.61981927  1.92833411]]
Boundary line intercepts are [  9.81609543   2.34942659 -12.16552202]


# Part 4: Support Vector Machine

In [6]:
# i. Use sklearn to train a Support Vector Classifier on the training set
clf_svm = make_pipeline(StandardScaler(), SVC(gamma="auto", probability=True))
clf_svm.fit(X=X_train, y=y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print("Probabilities for a random datapoint: ", clf_svm.predict_proba(X_test.sample()))


# iii. Report on the score for the SVM, what does the score measure?
print("Score:", clf_svm.score(X_test, y_test))
print("This score measures the mean accuracy on the given train data and labels. A number closer to 1 is more accurate.")



Probabilities for a random datapoint:  [[0.0077042  0.00292992 0.98936588]]
Score: 0.9333333333333333
This score measures the mean accuracy on the given train data and labels. A number closer to 1 is more accurate.


# Part 5: Neural Network

In [7]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(random_state=5, solver="adam")
clf_mlp = mlp.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print("Probabilities for a random datapoint: ", clf_mlp.predict_proba(X_test.sample()))

# iii. Report on the score for the Neural Network, what does the score measure?
print("Score (solver=\"adam\"):", clf_mlp.score(X_test, y_test))
print("This score measures the mean accuracy on the given train data and labels. A number closer to 1 is more accurate. I don't know why I have 1.0!!!")


# iv: Experiment with different options for the neural network, report on your best configuration
mlp = MLPClassifier(random_state=5, solver="lbfgs", max_iter=500)
clf_mlp = mlp.fit(X_train, y_train)
print("Score (solver=\"lbfgs\"): " , clf_mlp.score(X_train, y_train))

Probabilities for a random datapoint:  [[0.08360475 0.80537981 0.11101544]]
Score (solver="adam"): 1.0
This score measures the mean accuracy on the given train data and labels. A number closer to 1 is more accurate. I don't know why I have 1.0!!!
Score (solver="lbfgs"):  0.9925925925925926


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


# Part 6: K-Nearest Neighbors

In [26]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

neigh = KNeighborsClassifier(n_neighbors=6)
neigh.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
print("Probabilities for a random datapoint: ", neigh.predict_proba(X_test.sample()))
# iii. Report on the score for kNN, what does the score measure?
print("Score:", neigh.score(X_test, y_test))
print("This score measures the mean accuracy on the given train data and labels. A number closer to 1 is more accurate. I finally have a number that isn't 1 but close to it, yay!")



Probabilities for a random datapoint:  [[0. 0. 1.]]
Score: 0.9333333333333333
This score measures the mean accuracy on the given train data and labels. A number closer to 1 is more accurate. I finally have a number that isn't 1 but close to it, yay!


# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The results shown in each of the models suggest that all the models were very accurate in classifying the flowers. The neural network with the "adam" solver performed best on the dataset, with an accuracy of 1, while all the other models had the same accuracy (0.9333333333333333). This may be due to the fact that the neural netwrok was adapting learning rates during training, and allowed the neural network to converge quickly to an optimal solution. That's also what surprised me the most --- that one model had a perfect accuracy while all the other models, regardless of type or parameters, achieved the sasme accuracy. I guess that this may be because the dataset is relatively simple (owing to the fact that there are only 150 entries which is pretty small).